regression-tests/mixed-overview-of-as-casts.cpp2

```diff
@@ -0,0 +1,196 @@
+}
+
+struct ThrowingConstruction {
+    constexpr ThrowingConstruction() = default;
+    ThrowingConstruction(int) { throw 1; }
+};
+
+main: () = {
+    print_header();
+    { // nonesuch
+        i := 42;
+        print("(i as A) is (nonesuch)", "N/A", "N/A", "static assert - tested in separate tests");
+    }
+    {// smaller to bigger, bigger to smaller
+        print("(u8(12) as u16) is (u16(12))", (u8(12) as u16) is (u16(12)), true);
+        print("(u16(12) as u8) is (nonesuch)", "N/A", "N/A", "static assert - tested in separate tests");
+        print("(3.14f as double) is close_to(3.14f)", (3.14f as double) is (close_to(3.14f)), true, std::abs(3.14 - (3.14f as double)));
+        print("(3.14d as float) is close_to(3.14f)", "N/A", "N/A", "static assert - tested in separate tests");
+    }
+    {// signed/unsigned
+        print("( u8(12) as i16) is (i16(12))", ( u8(12) as i16) is (i16(12)), true);
+        print("(i16(12) as  u8) is ( u8(12))", "N/A", "N/A", "static assert - tested in separate tests");
+        print("( 12 as u8) is (u8(12))", (12 as u8) is (u8(12)), true);
+        print("( 12u as i8) is (i8(12))", (12u as i8) is (i8(12)), true);
+    }
+    {// integral to floating, floating to integral
+        print("( 12 as double) is (12.0)", (12 as double) is (12.0), true);
+        print("( 12.0 as int) is (12)", "N/A", "N/A", "static assert - tested in separate tests");
+    }
+    {// custom types casting
+        print("(X<12>() as int) is (int(X<12>()))", (X<12>() as int) is (int(X<12>())), true);
+    }
+    {
+        print("(3.14 as std::optional<int>) is (nonesuch)", "N/A", "N/A", "static assert - tested in separate tests");
+        print("(3 as std::optional<double>) is (std::optional<double>(3))", (3 as std::optional<double>) is (std::optional<double>(3)), true);
+    }
+    {// base_of
+        print("( A() as A ) is A", ( A() as A ) is A, true);
+        print("( C() as A ) is A", ( C() as A ) is A, true);
+        print("( B() as A ) is nonesuch_", "N/A", "N/A", "static assert - tested in separate tests");
+    }
+    {// Polymorphic types
+        vc: VC = ();
+        vp0 : *VA<0> = vc&;
+        vp1 : *VA<1> = vc&;
+
+        print("( vp0 as *VC ) is (vc&)", ( vp0 as *VC ) is (vc&), true);
+        print("( vp1 as *VC ) is (vc&)", ( vp1 as *VC ) is (vc&), true);
+        print("( vp0 as *VA<1> ) is (vp1)", ( vp0 as *VA<1> ) is (vp1), true);
+
+        print("( vp0* as VC )& is (vc&)", ( vp0* as VC )& is (vc&), true);
+        print("( vp1* as VC )& is (vc&)", ( vp1* as VC )& is (vc&), true);
+        print("( vp0* as VA<1> )& is (vp1)", ( vp0* as VA<1> )& is (vp1), true);
+
+        vd: VD = ();
+        vp2 : *VA<2> = vd&;
+        print("( vp2* as VC )", expect_throws(:() = ( vp2$* as VC ) is (cpp2::nonesuch);), "throws");
+
+    }
+    {// Variant
+        v : std::variant<int, long, float, double, std::string> = 42;
+
+        print("( v{42} as int ) is (42)", ( v as int ) is (42), true);
```

Comment:
Seems like `fmt` uses parentheses: <https://github.com/fmtlib/fmt/blob/44f3d8a77cd7e05e9da92d68635abdb4da626e9e/doc/api.rst#formatting-variants>.

Reply:
I used them before Herb introduced fmt... should I replace them with something else?

Reply:
No, I think it's OK.
What stood out to me was the braces.
It looked like initialization, but upon closer inspection, it's the held value.
I got curious as to what `fmt` does, and found out it's `variant(42)`.